custom_model_runner/datarobot_drum/drum/typeschema_validation.py

```diff
@@ -123,244 +219,145 @@ def convert(data):
-                            "Datatypes incorrect, expected data to have {}".format(self.values)
-                        )
-                else:
-                    if types[dtype]:
-                        validation_errors.append(
-                            "Datatypes incorrect, unexpected type {} found, expected {}".format(
-                                dtype, self.values
-                            )
-                        )
-        elif self.condition == Conditions.NOT_EQUALS:
-            if types[self.values[0]]:
-                validation_errors.append(
-                    "Datatypes incorrect, {} was expected to not be present".format(self.values)
-                )
-        elif self.condition == Conditions.NOT_IN:
-            for dtype in self.values:
-                if types[dtype]:
-                    validation_errors.append(
-                        "Datatypes incorrect, {} was expected to not be present".format(self.values)
-                    )
-
-        elif self.condition == Conditions.IN:
-            for dtype in self.VALUES:
-                if dtype in self.values:
-                    if not types[dtype]:
-                        validation_errors.append(
-                            "Datatypes incorrect, expected {} to be present".format(dtype)
-                        )
-                elif types[dtype]:
-                    validation_errors.append(
-                        "Datatypes incorrect, {} is not  expected to be present".format(dtype)
-                    )
-        return validation_errors
-
-
-class SparsityInput(BaseValidator):
-    FIELD = "sparse"
-    CONDITIONS = Conditions.EQUALS
-    VALUES = [Values.FORBIDDEN, Values.SUPPORTED, Values.REQUIRED]
+        types_present = [k for k, v in types.items() if v]
 
-    def validate(self, dataframe):
-        errors = []
-        if dataframe.dtypes.apply(pd.api.types.is_sparse).any():
-            if self.values not in [Values.SUPPORTED, Values.REQUIRED]:
-                errors.append(
-                    "Sparse input data found, however value is set to {}, expecting dense".format(
-                        self.values
-                    )
-                )
-        elif self.values not in [Values.FORBIDDEN, Values.SUPPORTED]:
-            errors.append("Dense input data found, however value is set to {}, expecting sparse")
-        return errors
+        base_error = f"Datatypes incorrect. Data has types: {types_present}"
+
+        errors = {
+            Conditions.EQUALS: f"{base_error}, but expected only {self.values[0]}.",
+            Conditions.NOT_EQUALS: f"{base_error}, but expected {self.values[0]} to NOT be present.",
+            Conditions.IN: f"{base_error}, but expected types to exactly match: {self.values}",
+            Conditions.NOT_IN: f"{base_error}, but expected no types in: {self.values} to be present",
+        }
 
+        tests = {
+            Conditions.EQUALS: lambda data_types: self.values == data_types,
```

Comment:
shouldn't this also check the length of data_types to match the expected behavior

Reply:
the check happens in BaseValidator.__init__;  any equals condition needs a values of `len == 1`.  if data_types is not _exactly_ a len==1 list, this will fail.